added a decorator function that allows nesting of Cmd classes to create submenus

lobocv opened the pull request:
Using the @AddSubmenu decorator, you can create a command that links you into a submenu (a Cmd class).

Please check out examples/submenus.py

This closes #256.

tleonhardt commented:
@lobocv 
Thank you for this PR Calvin!  From taking a brief look at it, I think it looks very good.  From an architectural viewpoint it looks solid and non-intrusive.  

I will take a detailed look at it this evening and/or over the weekend.  @kotfu If you have time to look at it as well, your feedback would be appreciated.

So far this looks like a nice PR.  But there are a few additional things which would make it a truly excellent PR:

1. pytest unit/integration test(s) for the new features added - you can use the same example or a pared-down version thereof if you wish to create integration tests
2. A section in the Sphinx documentation briefly describing how the new decorator works and linking to the nice example which is already included
3. A single line added under the "Main Features" in the README.md to let people know the feature exists

If you have time to add any and/or all of these, it would be appreciated.  None of these should involve an extensive amount of work.  If not, I'll take a stab at it.

Thanks again, I think the ability to easily create sub-menus is a great new feature ;-)

tleonhardt commented on cmd2.py:
```diff
@@ -438,6 +438,229 @@ def strip_ansi(text):
+        readline.add_history(line)
+
+
+def _complete_from_cmd(cmd, text, line, begidx, endidx):
```
cmd arg name shadows cmd module name from outer scope, best to be clear and rename argument

tleonhardt commented on cmd2.py:
```diff
@@ -438,6 +438,229 @@ def strip_ansi(text):
+
+        self.create_subclass = create_subclass
+
+    def __call__(self, Cmd):
```
Cmd arg name shadows cmd2.Cmd module name from outer scope, best to be clear and rename argument

tleonhardt commented on cmd2.py:
```diff
@@ -438,6 +438,229 @@ def strip_ansi(text):
+            tokens = line.split(None, 1)
+            if len(tokens) == 2 and (
+                tokens[1].startswith(self.command) or
+                any(tokens[1].startswith(alias) for alias in self.aliases)
```
alias local variable name shadows name from outer scope of __call__ implementation below, best to be clear and rename one of them

tleonhardt commented on cmd2.py:
```diff
@@ -438,6 +438,229 @@ def strip_ansi(text):
+
+
+class AddSubmenu(object):
+    """Conveniently add a submenu (Cmd-like class) to a Cmd
```
Nice job with good comments throughout.  Thank you.

tleonhardt commented on cmd2.py:
```diff
@@ -1188,6 +1411,7 @@ def do_shortcuts(self, args):
     def do_eof(self, arg):
         """Called when <Ctrl>-D is pressed."""
         # End of script should not exit app, but <Ctrl>-D should.
+        print('')  # Required for clearing line when exiting submenu
```
If you could find a clever way so that this conditionally executes based on some attribute which is only present within a submenu, that would be preferable just out of paranoia for it potentially having a negative effect on someones existing scripts and/or transcript tests. 

 If not, it seems like a pretty minor change which I don't think would be relevant for most end users.

tleonhardt commented on examples/submenus.py:
```diff
@@ -0,0 +1,90 @@
+"""
```
Minor recommendation, add a shebang line at the top and make it executable by user by default so it can be ran directly on macOS and Linux, i.e.:

```Python
#!/usr/bin/env python
```

tleonhardt commented on examples/submenus.py:
```diff
@@ -0,0 +1,90 @@
+Create a CLI with a nested command structure as follows. The commands 'second' and 'third' navigate the CLI to the scope
+of the submenu. Nesting of the submenus is done with the cmd2.AddSubmenu() decorator.
+
+      (Top Level)----second----->(2nd Level)----third----->(3rd Level)
```
Thank you for creating a nice example to go along with the submenu feature.  It goes a long way to explaining how to properly use it.

tleonhardt commented:
On a side note, it looks like your branch this PR was created from is a lot of commits behind master in the main repo, so just to be safe you should probably merge from the main repo's master branch into your fork; thought GitHub isn't giving any warnings so it is probably OK.

lobocv commented:
Wow, thanks for all your very valid comments. I will get to them all, I promise.
Do you have any suggestions on how to get the code coverage up?

tleonhardt commented:
Believe it or not, I think it should be relatively easy to get the code coverage back up.  The vast majority of our unit tests are actually relatively high level integration tests.  I'd recommend taking a look at our tests which were added recently to test the new sub-commands feature.  These are located in [test_argparse.py](https://github.com/python-cmd2/cmd2/blob/master/tests/test_argparse.py).  

The basic approach is as follows (all references are in the above file):

1. Create a class which inherits from ``cmd2.Cmd`` - see **SubcommandApp** for reference
    - In your case, this will probably be at least two classes, one serving as a sub-menu within the other

2. Create a pytest test fixture which creates an instance of this class and incorporates a StdOut hack which can then be passed to pytest unit tests - see **subcommand_app** for reference

3. Implement 1 to N pytest unit test functions which test the features you want using the above test fixture - see the six **test_subcommand_*** functions for reference

tleonhardt commented:
I went ahead and merged master into your fork/branch.  So make sure you pull before working on this again.

kotfu commented:
I had a chance to look at this earlier this morning. @lobocv, I agree with @tleonhardt, this is a nice bit of work, and I think it makes a great addition to cmd2.

lobocv commented:
Alright, hopefully I've addressed all your suggestions. Let me know if I've missed something.
I had a hard time writing tests for the submenu because of how the Stdout routing is set up. It actually by-passes the cmdloop which makes it hard to test entering submenus.
I added a few tests for calling the submenu commands from the top level but could not add tests of entering the submenu and calling commands.

tleonhardt commented:
I should have mentioned the ``pytest`` **capsys** fixture.  You can pass it in as a secondary test fixture and use it to capture stdout and stderr and then dont' need the StdOut hack.  See this for more info:
https://docs.pytest.org/en/2.9.0/capture.html

We use it in lots of tests which would otherwise be a real pain to write.

I'll take a detailed look at everything later tonight.  But right now it looks like one or more unit tests is failing on Python 2.7 only.

tleonhardt commented:
@lobocv I think for this one you are going to want a 2nd StdOut hack placed within the second level Cmd2 instance.  And you will need to test both levels.

The start of the unit tests might look something like the following:
```Python
import pytest

import cmd2
from conftest import run_cmd, StdOut, normalize


class SecondLevel(cmd2.Cmd):
    """To be used as a second level command class. """
    def __init__(self, *args, **kwargs):
        cmd2.Cmd.__init__(self, *args, **kwargs)
        self.prompt = '2ndLevel '
        self.top_level_attr = None

    def do_say(self, line):
        self.poutput("You called a command in SecondLevel with '%s'. " % line)

    def help_say(self):
        self.poutput("This is a second level menu. Options are qwe, asd, zxc")

    def complete_say(self, text, line, begidx, endidx):
        return [s for s in ['qwe', 'asd', 'zxc'] if s.startswith(text)]

    def do_get_top_level_attr(self, line):
        self.poutput(str(self.top_level_attr))

    def do_get_prompt(self, line):
        self.poutput(self.prompt)


second_level_cmd = SecondLevel()
second_out = StdOut()
second_level_cmd.stdout = second_out


@cmd2.AddSubmenu(second_level_cmd,
                 command='second',
                 aliases=('second_alias',),
                 shared_attributes=dict(top_level_attr='top_level_attr'))
class SubmenuApp(cmd2.Cmd):
    """To be used as the main / top level command class that will contain other submenus."""

    def __init__(self, *args, **kwargs):
        cmd2.Cmd.__init__(self, *args, **kwargs)
        self.prompt = 'TopLevel '
        self.top_level_attr = 123456789

    def do_say(self, line):
        self.poutput("You called a command in TopLevel with '%s'. " % line)

    def help_say(self):
        self.poutput("This is a top level submenu. Options are qwe, asd, zxc")

    def complete_say(self, text, line, begidx, endidx):
        return [s for s in ['qwe', 'asd', 'zxc'] if s.startswith(text)]


@pytest.fixture
def submenu_app():
    app = SubmenuApp()
    app.stdout = StdOut()
    return app


def test_submenu_say_from_top_level(submenu_app):
    line = 'testing'
    out = run_cmd(submenu_app, 'say ' + line)
    assert len(out) == 1
    assert out[0] == "You called a command in TopLevel with {!r}.".format(line)

def test_submenu_second_say_from_top_level(submenu_app):
    line = 'testing'
    out = run_cmd(submenu_app, 'second say ' + line)

    # No output expected from the top level
    assert out == []

    # Output expected from the second level
    out2 = normalize(second_out.buffer)
    second_out.clear()
    assert len(out2) == 1
    assert out2[0] == "You called a command in SecondLevel with {!r}.".format(line)
```

tleonhardt commented on tests/test_submenu.py:
```diff
@@ -0,0 +1,107 @@
+    second_level_cmd.stdout = app.stdout = StdOut()
+    return app
+
+
```
Recommend adding a single happy-day test of top-level behavior, such as:

```Python
def test_submenu_say_from_top_level(submenu_app):
    line = 'testing'
    out1, out2 = run_submenu_cmd(submenu_app, 'say ' + line)
    assert len(out1) == 1
    assert len(out2) == 0
    assert out1[0] == "You called a command in TopLevel with {!r}.".format(line)
```

tleonhardt commented on tests/test_submenu.py:
```diff
@@ -0,0 +1,107 @@
+    return app
+
+
+def test_second_say_from_top_level(submenu_app):
```
Recommend beginning all tests names in this file with **test_submenu_*** to assist with py.test keyword matching so you can run them from the main directory using:

```Bash
py.test -vk submenu
```

tleonhardt commented on tests/test_submenu.py:
```diff
@@ -0,0 +1,107 @@
+
+
+def test_second_say_from_top_level(submenu_app):
+    line = 'testing'
```
Using the **run_submenu_cmd** convenience function, this test can be modified to something like the following:

```Python
def test_submenu_second_say_from_top_level(submenu_app):
    line = 'testing'
    out1, out2 = run_submenu_cmd(submenu_app, 'second say ' + line)

    # No output expected from the top level
    assert out1 == []

    # Output expected from the second level
    assert len(out2) == 1
    assert out2[0] == "You called a command in SecondLevel with {!r}.".format(line)
```

Hopefully this can serve as a guide for the remaining tests.

tleonhardt commented on tests/test_submenu.py:
```diff
@@ -0,0 +1,107 @@
+    assert out == [str(submenu_app.top_level_attr)]
+
+
+if __name__ == '__main__':
```
py.test keyword matching along with a consistent naming prefix makes this __main__ section unnecessary.  It doesn't hurt anything though.

tleonhardt commented:
@lobocv Since I am very familiar with the idiosyncrasies of unit/integration testing features for ``cmd2``, I am working on pushing some unit test fixes to your fork for this PR.  This is perhaps the most painful feature I have seen to test due to requiring multiple different StdOut interceptions from different locations/levels.

tleonhardt commented on cmd2.py:
```diff
@@ -547,6 +547,230 @@ def strip_ansi(text):
+                    submenu.precmd(line)
+                    ret = submenu.onecmd(line)
+                    submenu.postcmd(ret, line)
+                else:
```
I'd like to try to find a unit/integration test which covers this **else** case to help increase code coverage.  Can you suggest one?

tleonhardt commented:
I'm going to merge this PR in.  The unit tests for this new sub-menu feature could use some improvement, but that doesn't need to happen as part of this PR.  It can be done as follow-on work as time permits.
